rditer,kvserver: simplify and unify replica span handling

tbg opened the pull request:
From 40573d34146328a0f218c5bb9a290a6e5ee3ee55 Mon Sep 17 00:00:00 2001
From: Tobias Grieger <tobias.b.grieger@gmail.com>
Date: Fri, 6 Jan 2023 05:42:50 +0100
Subject: [PATCH] rditer,kvserver: simplify and unify replica span handling

A Replica comprises a number of spans:

- RangeID-keyed replicated data, such as the `GCThresholdKey`,
  confusingly named "local RangeID keys" since they all have a `/Local/`
(`\x02`) prefix. They have a `r` infix.
- RangeID-keyed unreplicated data, such as the `HardState`,
- The system-internal key-addressable "parallel plane" (such as the
  `RangeDescriptorKey`), confusingly named "local keys" (again, `\x02`
prefix, but `u` infix).
- Two lock spans: one for the "local" keyspace (e.g. for an intent on
  a `RangeDescriptor`), one for the "global" keyspace (e.g. regular
  row intents), in both cases the span is local and uses a `k` infix.
- The actual span of the range (i.e. user data).

Code that needs some or all of these spans gets them from various
methods provided by the `rditer` package. Other than being generally a
messy affair, with a separate raft log, RangeID-based unreplicated keys
will mostly[^1] live on the logstore engine, while RangeID-based
replicated keys will be on the state machine engine and so the current
approach of targeting either only the replicated *or* the combined
(replicated and unreplicated) part of the local RangeID-based span
no longer works. It was also never particularly obvious which spans
one would select with that bool, and I wound up confused more than
once.

The above means we need to stop thinking about RangeID-based keys as
belonging to "one span" - the two RangeID-keyed spans are very different
in nature and storage.

This commit introduces a unified way to select the different parts of the
keyspace. It then goes ahead and uses this new primitive for all span
uses external to the `rditer` package. (`rditer` itself has a few methods
that need exotic combinations of spans, and it gets to do those more
ad-hoc for now but this could be improved later).

In particular, we can now rework `clearRangeData` and give it an
analogoous way to explicitly select which spans to clear. Replica
removal is a central work item in CRDB-220, since it comes up all
over the place: snapshots (in particular in combination with replica
subsumption), splits, replication changes, replicaGC.

After this PR, we can improve splits by not mutating the raft state of
the right-hand side replica in cases where it is known to be newer than
the split. The old API forced us to remove the raft state as well, and
we had to awkwardly lock the right-hand side and write the raft state
back to make things work. We can now simply opt out of mutating that
raft state (via `ClearUnreplicatedByRangeID=false`). This is left for
a separate PR.

This PR makes some additional improvements:

- improve `rditer` tests via `echotest`, also test lock table keys
- improved commentary, esp. around [^1].

[^1]: https://github.com/cockroachdb/cockroach/issues/95055

Release note: None
Epic: CRDB-220

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/94201)
<!-- Reviewable:end -->

tbg committed:
rditer,kvserver: simplify and unify replica span handling

A Replica comprises a number of spans:

- RangeID-keyed replicated data, such as the `GCThresholdKey`,
  confusingly named "local RangeID keys" since they all have a `/Local/`
(`\x02`) prefix. They have a `r` infix.
- RangeID-keyed unreplicated data, such as the `HardState`,
- The system-internal key-addressable "parallel plane" (such as the
  `RangeDescriptorKey`), confusingly named "local keys" (again, `\x02`
prefix, but `u` infix).
- Two lock spans: one for the "local" keyspace (e.g. for an intent on
  a `RangeDescriptor`), one for the "global" keyspace (e.g. regular
  row intents), in both cases the span is local and uses a `k` infix.
- The actual span of the range (i.e. user data).

Code that needs some or all of these spans gets them from various
methods provided by the `rditer` package. Other than being generally a
messy affair, with a separate raft log, RangeID-based unreplicated keys
will mostly[^1] live on the logstore engine, while RangeID-based
replicated keys will be on the state machine engine and so the current
approach of targeting either only the replicated *or* the combined
(replicated and unreplicated) part of the local RangeID-based span
no longer works. It was also never particularly obvious which spans
one would select with that bool, and I wound up confused more than
once.

The above means we need to stop thinking about RangeID-based keys as
belonging to "one span" - the two RangeID-keyed spans are very different
in nature and storage.

This commit introduces a unified way to select the different parts of the
keyspace. It then goes ahead and uses this new primitive for all span
uses external to the `rditer` package. (`rditer` itself has a few methods
that need exotic combinations of spans, and it gets to do those more
ad-hoc for now but this could be improved later).

In particular, we can now rework `clearRangeData` and give it an
analogoous way to explicitly select which spans to clear. Replica
removal is a central work item in CRDB-220, since it comes up all
over the place: snapshots (in particular in combination with replica
subsumption), splits, replication changes, replicaGC.

After this PR, we can improve splits by not mutating the raft state of
the right-hand side replica in cases where it is known to be newer than
the split. The old API forced us to remove the raft state as well, and
we had to awkwardly lock the right-hand side and write the raft state
back to make things work. We can now simply opt out of mutating that
raft state (via `ClearUnreplicatedByRangeID=false`). This is left for
a separate PR.

This PR makes some additional improvements:

- improve `rditer` tests via `echotest`, also test lock table keys
- improved commentary, esp. around [^1]

[^1]: https://github.com/cockroachdb/cockroach/issues/95055

Release note: None

tbg commented:
I suggest starting the review with two windows, one looking at old diff for `makeRangeKeySpans` and comparing it to `Select` in new file `select.go`. The logic essentially transitioned over.

The remainder of the changes in `rditer` remove custom implementations and use `Select` instead:

- `MakeReplicatedKeySpansExceptRangeID` was removed, since we could just use `Select` directly.
- `MakeRangeIDLocalKeySpan` is no longer there because we now have two methods to make spans for the replicated and unreplicated parts, respectively: `makeRangeID{Unre,Re}plicatedSpan`.

In `kvserver`, the main change affects `clearRangeData`, which wants to use the new API:

- `clearRangeDataOptions` is introduced and basically mirrors the options to `Select`
- all callers are updated to use it.

tbg committed:
echotest: create directory when --rewrite passed

Epic: none
Release note: None

pavelkalinnikov commented on pkg/kv/kvserver/store_split.go:
```diff
@@ -105,9 +85,24 @@ func splitPreApply(
+		if err := clearRangeData(split.RightDesc.RangeID, readWriter, readWriter, clearRangeDataOptions{
+			// We know there isn't anything in these two replicated spans below in the
+			// right-hand side (before the current batch), so setting these options
+			// will in effect only mask the writes to the RHS replicated state we have
+			// staged in this batch, which is what we're after.
```
What does "mask" mean here? Will it remove those staged writes?

tbg replied:
"clear",  fixed.

pavelkalinnikov commented on pkg/kv/kvserver/rditer/select.go:
```diff
@@ -0,0 +1,104 @@
+
+import "github.com/cockroachdb/cockroach/pkg/roachpb"
+
+type StateMachineSelectionOptions struct {
```
Do you plan to use this struct in isolation from `SelectionOptions`? If not, I would suggest inlining it, to avoid the verbosity that it carries (long name repetitions + extra nesting on every `Select` call; e.g. see calls in `replica_raftstorage.go`).

tbg replied:
Good idea, this was unnecessarily verbose. Done.

pavelkalinnikov commented on pkg/kv/kvserver/replica_raftstorage.go:
```diff
@@ -792,7 +804,16 @@ func (r *Replica) clearSubsumedReplicaDiskData(
+		return rditer.Select(d.RangeID, rditer.SelectionOptions{
+			StateMachineSelectionOptions: rditer.StateMachineSelectionOptions{
+				ReplicatedBySpan: roachpb.RSpan{
+					Key:    d.StartKey,
+					EndKey: d.EndKey,
```
Is there any use-case where `StartKey`/`EndKey` is not taken from the range descriptor? If so, directly passing the descriptor (instead of taking range ID + span manually) can be considered "safer".

tbg replied:
It feels wrong to use a `RangeDescriptor` here. That is a very large struct. But I was able to replace this with just `d.RSpan()`.

pavelkalinnikov commented on pkg/kv/kvserver/rditer/testdata/TestSelect/replicatedid_span:
```diff
@@ -0,0 +1,17 @@
+StateMachineSpans():
+- /Local/RangeID/123/{r""-s""}
+- /Local/Range"{a"-c"}
+- /Local/Lock/Intent/Local/Range"{a"-c"}
+- /Local/Lock/Intent"{a"-c"}
+- /Local/Lock/Intent/Local/Range"{a"-c"}
+- /Local/Lock/Intent"{a"-c"}
```
Why are these duplicated?

tbg replied:
Because there was a bug - I was appending to the wrong slice in one place in `Select`. Thanks for catching this, I should've found this myself.

pavelkalinnikov commented on pkg/kv/kvserver/rditer/testdata/TestSelect/replicatedid_unreplicatedid_span:
```diff
@@ -0,0 +1,20 @@
+- /Local/Lock/Intent"{a"-c"}
+- {a-c}
+NonStateMachineSpans():
+- /Local/RangeID/123/{u""-v""}
```
Why is this span in both state and non-state-machine?

pavelkalinnikov commented on pkg/kv/kvserver/rditer/testdata/TestSelect/span:
```diff
@@ -0,0 +1,15 @@
+echo
+----
+Spans():
+- /Local/Range"{a"-c"}
+- /Local/Lock/Intent/Local/Range"{a"-c"}
+- /Local/Lock/Intent"{a"-c"}
+- {a-c}
+StateMachineSpans():
+- /Local/Range"{a"-c"}
+- /Local/Lock/Intent/Local/Range"{a"-c"}
+- /Local/Lock/Intent"{a"-c"}
+- /Local/Lock/Intent/Local/Range"{a"-c"}
+- /Local/Lock/Intent"{a"-c"}
+- {a-c}
```
Why is the "total" list shorter than the state-machine-only spans list?

pav-kv reviewed (commented):
Dumping some comments for today.

tbg commented on pkg/kv/kvserver/rditer/select.go:
```diff
@@ -0,0 +1,104 @@
+// Select creates a Selection according to the SelectionOptions.
+func Select(rangeID roachpb.RangeID, opts SelectionOptions) Selection {
+	var s Selection
+
```
It's not hot, we only use this in paths like snapshot application, etc, where the overhead of allocating the slices automatically is negligible.

tbg commented on pkg/kv/kvserver/rditer/select.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+// Select creates a Selection according to the SelectionOptions.
+func Select(rangeID roachpb.RangeID, opts SelectionOptions) Selection {
```
👍🏽

tbg commented on pkg/kv/kvserver/rditer/select.go:
```diff
@@ -0,0 +1,104 @@
+// to read or erase only certain components of a Replica.
+type SelectionOptions struct {
+	StateMachineSelectionOptions
+	// ReplicatedByRangeID selects all RangeID-keyed unreplicated keys. Examples
```
Thanks, I miss the linter that points these out.

tbg commented on pkg/kv/kvserver/replica_raftstorage.go:
```diff
@@ -421,31 +421,43 @@ func (r *Replica) updateRangeInfo(ctx context.Context, desc *roachpb.RangeDescri
+	// to write additional keys to an SST after this clear, e.g. a replica
+	// tombstone, since keys must be written in order. When this is false, a
+	// heuristic will be used instead.
+	MustUseClearRange bool
```
I think the confusion is that the method is named `clearRangeData`. The option is named adequately: when set, we use `storage.ClearRangeWithHeuristic` to actually remove the data. This is not a MVCC Range Deletion Tombstone, so I would steer clear of anything that might suggest that.
Instead of making a change now, I added a TODO to rename `clearRangeData`:

> // TODO(tbg): could rename this to clearReplicaData. The use of "Range" in both the
// "CRDB Range" and "storage.ClearRange" context in the setting of this method could
// be confusing.

tbg commented on pkg/kv/kvserver/store_split.go:
```diff
@@ -105,9 +85,24 @@ func splitPreApply(
+			//
+			// TODO(tbg): we don't actually want to touch the raft state of the right
+			// hand side replica since it's absent or a more recent replica than the
+			// split. Now that we have a boolean targeting the unreplicated
+			// RangeID-based keyspace, we can set this to false and remove the
```
We shouldn't touch the RHS raft state here at all. The only reason we do this writeback is because the previous API forced us to remove everything. The RHS is for all intents and purposes a completely unrelated replica in this context, so there shouldn't be a reason to touch it.
But you have a good point that it's not as simple as that. We are operating here on the `WriteBatch` that the split is using. We really are trying to remove writes from that WriteBatch, rather than trying to remove anything from the engine. But there isn't a way to "delete just from the batch but not from the engine the batch sits on top of". So the question is, what is in the `WriteBatch`?

The relevant comment is here:

https://github.com/cockroachdb/cockroach/blob/f263a765d750e41f2701da0a923a6e92d09159fa/pkg/kv/kvserver/batcheval/cmd_end_transaction.go#L1109-L1149

The long and short of it is that we don't write to the RangeID-unreplicated key space in the WriteBatch - we rely on the call to `SynthesizeRaftState` (which only happens further down below, and which we won't reach in this case since we're returning early).

Expanded the comment a bit with this information.

tbg commented on pkg/kv/kvserver/store_split.go:
```diff
@@ -69,26 +69,6 @@ func splitPreApply(
 		// LHS due to the split and additionally clear all of the range ID local state
 		// that the split trigger writes into the RHS.
-		//
-		// We know we've never processed a snapshot for the right range because the
-		// LHS prevents any incoming snapshots until the split has executed (i.e. now).
-		// It is important to preserve the HardState because we might however have
-		// already voted at a higher term. In general this shouldn't happen because
-		// we add learners and then promote them only after we snapshot but we're
-		// going to be extra careful in case future versions of cockroach somehow
-		// promote replicas without ensuring that a snapshot has been received.
```
Good point, restored this in a modified form.

tbg commented on pkg/kv/kvserver/rditer/replica_data_iter.go:
```diff
@@ -54,104 +54,79 @@ type ReplicaMVCCDataIterator struct {
 
-// MakeAllKeySpans returns all key spans for the given Range, in
+// makeAllKeySpans returns all key spans for the given Range, in
 // sorted order.
```
Good idea, done.

tbg commented on pkg/kv/kvserver/rditer/select.go:
```diff
@@ -0,0 +1,104 @@
+// in the tests for this package that attempts to create an interesting data set,
+// but we should "somehow" make sure this data set is always complete, i.e. contains
+// all possible kinds of keys even as new ones are being added.
+type Selection struct {
```
You will need to clear both statemachine and logstorage state, and they live on different engines. So you want to have two lists of spans. But, you are right, I could call `Select` twice and select separately. I like that, it's simpler. I gave that a shot, PTAL.

pav-kv reviewed (commented):
Commented the fixups. Looking good. I'll make a final pass on the whole PR after this little round later today.

pavelkalinnikov commented on pkg/kv/kvserver/rditer/select.go:
```diff
@@ -30,66 +30,38 @@ type SelectOpts struct {
 	}
 
 	if in := opts.ReplicatedBySpan; !in.Equal(roachpb.RSpan{}) {
 		d := &roachpb.RangeDescriptor{StartKey: in.Key, EndKey: in.EndKey}
```
Also, right here it could be created without the `&`. Don't know if it changes allocation, but communicates the intention to not allocate.

tbg replied:
Added a commit that avoids need for RangeDescriptor.

tbg commented on pkg/kv/kvserver/rditer/select_test.go:
```diff
@@ -76,17 +76,8 @@ func TestSelect(t *testing.T) {
-			fmt.Fprintln(&buf, "NonStateMachineSpans():")
-			for _, sp := range tc.s.NonStateMachineSpans() {
-				fmt.Fprintf(&buf, "- %s\n", sp)
+			for _, sp := range tc.s {
+				fmt.Fprintf(&buf, "%s\n", sp)
```
Revamped this, see new commit. Should be pretty vanilla now! Thanks for the nudge, much better.

tbg commented on pkg/kv/kvserver/rditer/select.go:
```diff
@@ -31,9 +31,8 @@ type SelectOpts struct {
-// Which fields are populated depends on the SelectOpts passed to
-// Select.
+// keyspace, depending on the supplied SelectOpts. The resulting Spans are ordered
+// in strictly ascending lexicographical (`bytes.Compare`) order.
```
It is now!

tbg replied:
(Though I think the rditer tests verify it already, indirectly)

pavelkalinnikov commented on pkg/kv/kvserver/rditer/select_test.go:
```diff
@@ -0,0 +1,76 @@
+				Key:    roachpb.RKey("a"),
+				EndKey: roachpb.RKey("c"),
+			},
+		},
```
How about also a test with a span going to the end of the keyspace? Would it differ in any way from `r2`?

tbg replied:
It shouldn't! I added it to `TestSelect` and it performs as expected.

tbg commented on pkg/kv/kvserver/rditer/select_test.go:
```diff
@@ -24,60 +24,57 @@ import (
+					if len(description) == 0 {
+						description = append(description, "nothing")
+					}
+					fmt.Fprintf(&buf, "Select(%s):\n", strings.Join(description, ", "))
```
Hitting the like button.

tbg commented:
Could I trouble you for a last review? I just fixed a pretty severe (but at least caught in 100 places across CI) problem, see the new comits.

tbg commented:
Friendly ping.

pavelkalinnikov commented on pkg/kv/kvserver/rditer/select.go:
```diff
@@ -0,0 +1,85 @@
+		// user keys we have to.
+		//
+		// See also the comment on KeySpan.
+		in := opts.ReplicatedBySpan
```
nit: embed this to the `if` line

pavelkalinnikov commented on pkg/kv/kvserver/rditer/select.go:
```diff
@@ -0,0 +1,85 @@
+
+		// Lock table.
+		{
+			// Handle doubly-local lock table keys since range descriptor key
+			// is a range local key that can have a replicated lock acquired on it.
+			startRangeLocal, _ := keys.LockTableSingleKey(keys.MakeRangeKeyPrefix(in.Key), nil)
+			endRangeLocal, _ := keys.LockTableSingleKey(keys.MakeRangeKeyPrefix(in.EndKey), nil)
+			// Need .KeySpan() to avoid overlapping with the local lock span right above.
+			startGlobal, _ := keys.LockTableSingleKey(in.KeySpan().Key.AsRawKey(), nil)
+			endGlobal, _ := keys.LockTableSingleKey(in.KeySpan().EndKey.AsRawKey(), nil)
+			sl = append(sl, roachpb.Span{
+				Key:    startRangeLocal,
+				EndKey: endRangeLocal,
+			}, roachpb.Span{
+				Key:    startGlobal,
+				EndKey: endGlobal,
+			})
```
I hardly understand this block, hopefully you've verified/tested it.

pavelkalinnikov commented on pkg/kv/kvserver/rditer/select.go:
```diff
@@ -0,0 +1,85 @@
+			})
+		}
+		// Call KeySpan() because r1's "normal" keyspace starts only at LocalMax, not RKeyMin.
+		sl = append(sl, in.KeySpan().AsRawSpanWithNoLocals())
```
`in.KeySpan()` is repeated 3 times, and also has a conditional change in it. Consider saving it to a variable and reusing.

pavelkalinnikov commented on pkg/kv/kvserver/replicate_queue_test.go:
```diff
@@ -189,24 +191,48 @@ func TestReplicateQueueRebalanceMultiStore(t *testing.T) {
 	allocatorimpl.LeaseRebalanceThreshold = 0.01
 	allocatorimpl.LeaseRebalanceThresholdMin = 0.0
 
+	const useDisk = false // for debugging purposes
```
Is this needed only for this test? Could this be lifted to the harness?

tbg commented:
TFTR! I skipped the few nits you had to get this PR merged, but I did make the non-nit adjustment.
bors r=pavelkalinnikov

tbg commented:
>        	Error:      	"change replicas of r54 failed: descriptor changed: [expected] r54:/T{able/53-enant/10} [(n1,s1):1, (n2,s2):2, (n5,s5):3, (n3,s3):5NON_VOTER, (n4,s4):6NON_VOTER, next=7, gen=9] != [actual] r54:/T{able/53-enant/10} [(n1,s1):1, (n2,s2):2, (n5,s5):3NON_VOTER, (n3,s3):5, next=7, gen=12]" does not contain "ok"

https://github.com/cockroachdb/cockroach/issues/95262

bors r=pavelkalinnikov
